orbis-kernel/include/orbis/utils/SharedMutex.hpp

```diff
@@ -101,6 +104,11 @@ struct shared_mutex final {
 
   // Check whether can immediately obtain a shared (reader) lock
   bool is_lockable() const { return m_value.load() < c_one - 1; }
+#if defined(MAC)
+private:
+  pthread_mutex_t m_mutex;  // Mutex for synchronization
+  pthread_cond_t m_cond;    // Condition variable for signaling
+#endif
```

Comment:
Mutex should only occupy small zero-initialized area of 4 bytes, otherwise it causes problems. Also, these primitives should reliably perform in multi-process environment. If you are really interested in porting this, I can recommend to take a look at futex emulation in RPCS3 for inspiration. https://github.com/RPCS3/rpcs3/blob/master/Utilities/sync.h